### ISSUE TITLE
Remove usage of typeid, introduce type traits

### ISSUE THREAD

**Issue by tamiko** [dealii/dealii#3597]
A number of places in the library use constructs of the form
(example `include/deal.II/lac/constraint_matrix.templates.h`)
```
 856   // FIXME: This has to be refactored into a typetrait                           
 857   if ((typeid(vec) != typeid(Vector<double>)) &&                                 
 858       (typeid(vec) != typeid(Vector<float>)) &&                                  
 859       (typeid(vec) != typeid(Vector<std::complex<double> >)) &&                  
 860       (typeid(vec) != typeid(BlockVector<double>)) &&                            
 861       (typeid(vec) != typeid(BlockVector<float>)) &&                             
 862       (typeid(vec) != typeid(BlockVector<std::complex<double> >)) &&             
 863       (typeid(vec) != typeid(LinearAlgebra::Vector<double>)) &&                  
 864       (typeid(vec) != typeid(LinearAlgebra::Vector<float>)) &&                   
 865       (typeid(vec) != typeid(LinearAlgebra::Vector<std::complex<double> >)))
```

All of these instances should be refactored into proper typetraits, e.g.
```
if(is_basic_vector<...>::value) {
  ...
}
```

In reference to #3592

**Comment by koecher** [comment]
Dear @tamiko can you be a bit more specific, where do you find those. Searching in all deal.II files can be a bit longish.

**Comment by bangerth** [comment]
@koecher -- is this fixed?

**Comment by koecher** [comment]
@bangerth this issue is resolved by PR #3784 
I've searched through the current developer version of deal.II with:
`find . -type f -exec grep -H 'typeid(' {} \;`
and do not find anymore places regarding this issue.
Also the `static const bool supports_distributed_data` fields within all vector declarations are not used within the library. (They could be deprecated now).

**Comment by bangerth** [comment]
OK, let's close the current issue then. @koecher -- can you open another issue for the deprecation?
